README.md
# Five.py

[![Latest Version][1]][2]
[![Build Status][3]][4]
[![Coverage Status][5]][6]

A library to overcomplicate 5. It's a python port of the javascript [five][].


## Install

    $ pip install five


## Usage

```python
from five import Five
five = Five()
```

### Basic 5
```python
five()  # 5
five() + five()  # 10
five() * five()  # 25
five() / five()  # 1
five() - five()  # 0
five.pow(3)      # 125
five.sqrt()      # 2.23606797749979
```

### Counting is simple
```python
five() + five()  # 10
five() + five() - five() / five()  # 9
five() + (five() + five() + five()) / five()  # 8
five() + (five() + five()) / five()  # 7
five() + five() / five()  # 6
five()  # 5
five() - five() / five()  # 4
five() - (five() + five()) / five()  # 3
(five() + five()) / five()  # 2
five() / five()  # 1
```

### Math
```python
five.fibonacci()  # 5
five.factorial()  # 120
five.factors()  # [1, 5]
five.is_prime()  # True
```

### Different sorts of 5
```python
five.up_high()  # ⁵
five.down_low()  # ₅
five.too_slow()  # 5, with a 555-millisecond delay
five.roman()  # V
five.morse_code()  # .....(di-di-di-di-di)
five.negative()  # -5
five.loud()  # FIVE
five.loud('piglatin')  # IVEFAY
five.smooth()  # S
five.figlet()  #   _____
                  | ____|
                  | |__
                  |___ \
                   ___) |
                  |____/
five.stars()  # *****
five.bool()   # True
five.elements()  # ["Metal", "Wood", "Water", "Fire", "Earth"]
five.senses()  # ["Sight", "Hearing", "Taste", "Smell", "Touch"]
five.circled()  # ⑤
five.flipside()  # ϛ
five.lcd()  #  ---
              |
               --
                 |
              ---
```

### 5 goes multilingual
```python
five.afrikaans()  # vyf
five.armenian()  # հինգ
five.arabic()  # خمسة
five.azerbaijani()  # beş
five.basque()  # bost
five.belarusian  # пяць
five.bosnian()  # pet
five.bulgarian()  # пет
five.canadian()  # five eh
five.catalan()  # cinc
five.chinese()  # 五
five.chinese('pinyin')  # 'wǔ'
five.chinese('financial')  # '伍'
five.choctaw()  # tahlapi
five.croatian()  # pet
five.czech()  # pět
five.danish()  # fem
five.dovah()  # hen
five.dutch()  # vijf
five.elvish()  # lempe
five.english()  # Five
five.estonian()  # viis
five.finnish()  # viisi
five.french()  # cinq
five.georgian()  # ხუთი
five.german()  # fünf
five.greek()  # πέντε
five.hebrew()  # חמש
five.hindi()  # पांच
five.hungarian()  # öt
five.icelandic()  # fimm
five.indonesian()  # lima
five.irish()  # cúig
five.italian()  # cinque
five.japanese()  # 五
five.kannada()  # ಐದು
five.klingon()  # vagh
five.korean()  # 오
five.latin()  # quinque
five.latvian()  # pieci
five.lithuanian()  # penki
five.lojban()  # mu
five.manx()  # queig
five.mongolian()  # таван
five.norwegian()  # fem
five.persian()  # پنج
five.piglatin()  # ivefay
five.polish()  # pięć
five.portuguese()  # cinco
five.punjabi()  # ਪੰਜ
five.romanian()  # cinci
five.russian()  # пять
five.serbian()  # pet
five.slovakian()  # päť
five.slovenian()  # pet
five.spanish()  # cinco
five.swedish()  # fem
five.tagalog()  # lima
five.tamil()  # ஐந்து
five.telugu()  # ఐదు
five.thai()  # ห้า
five.turkish()  # beş
five.xhosa()  # zintlanu
five.ukrainian()  # п’ять
five.urdu()  # پانچ
five.vietnamese()  # năm
five.welsh()  # pump
```

### Different radices
```python
five.binary()  # 101
five.octal()  # 5
five.hex()  # 5
```

### Map and Reduce
```python
five.map([1, 2, 3])  # [5, 5, 5]
five.reduce([1, 2, 3])  # 5
```

### Novelty
```python
five.fab()  # ['Juwan Howard','Ray Jackson','Jimmy King','Jalen Rose','Chris Webber']
five.jackson()  # ['Jackie','Tito','Jermaine','Marlon','Michael']
five.famous_five()  # ['Julian', 'Dick', 'Anne', 'George', 'Timmy']
```

### Rotation
```python
five.rotate("five.py")  #knaj.ud
```

### Unicode
```python
five.oclock()  # '🕔'
```

## Development

run the tests:

    $ pip install -r dev-requirements.txt
    $ make test

## Credits

All the glories should belong to [@jackdcrawford][], I just port it to python :)

## License

MIT.

[1]: http://img.shields.io/pypi/v/five.svg
[2]: https://pypi.python.org/pypi/five
[3]: https://travis-ci.org/lord63/five.py.svg
[4]: https://travis-ci.org/lord63/five.py
[5]: https://coveralls.io/repos/lord63/five.py/badge.svg
[6]: https://coveralls.io/r/lord63/five.py
[five]: https://github.com/jackdcrawford/five
[@jackdcrawford]: https://github.com/jackdcrawford


ChangeLog
2015.06.14 v0.4.0
- add more lovely staffs
    - five.armenian()
    - five.circled()
    - five.factors()
    - five.flipside()
    - five.georgian()
    - five.is_prime()
    - five.lcd()
- bugfix for five.morse_code()
- update five.chinese()

2015.06.13 v0.3.0
- add more funny staffs
    - five.afrikaans()
    - five.bool()
    - five.canadian()
    - five.danish()
    - five.elements()
    - five.facotrial()
    - five.famous_five()
    - five.fibonacci()
    - five.lojban()
    - five.manx()
    - five.punjabi()
    - five.senses()
    - five.tagalog()
    - fie.urdu()
    - five.vietnamese()
    - five.xhosa()

2015.06.12 v0.2.0
- add more features:
    - five.icelandic()
    - five.pow()
    - five.sqrt()
    - five.figlet()
    - five.stars()

2015.05.23 v0.1.2
- initial port of the original javascript `five`.


five/test/test_five.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import absolute_import, unicode_literals, division

import math
import string
from unittest import TestCase

from five import Five
import pytest


class TestFive(TestCase):
    def setUp(self):
        self.five = Five()

    def test_basic_five(self):
        assert self.five() == 5
        assert self.five() != 6

    def test_operations_of_five(self):
        assert self.five() * self.five() == 25
        assert self.five() + self.five() == 10
        assert self.five() / self.five() == 1
        assert self.five() - self.five() == 0
        assert self.five() / self.five() * self.five() == self.five()
        assert self.five.pow(2) == 25
        assert self.five.sqrt() == math.sqrt(5)

    def test_math(self):
        assert self.five.fibonacci() == 5
        assert self.five.factorial() == 120
        assert self.five.factors() == [1, 5]
        assert self.five.is_prime() == True

    def test_different_sorts_of_five(self):
        assert self.five.up_high() == '⁵'
        assert self.five.down_low() == '₅'
        assert self.five.too_slow() == 5
        assert self.five.roman() == 'V'
        assert self.five.morse_code() == '.....'
        assert self.five.negative() == -5
        assert self.five.loud() == 'FIVE'
        assert self.five.loud('piglatin') == 'IVEFAY'
        assert self.five.smooth() == 'S'
        assert self.five.figlet() == (' _____\n| ____|\n| |__\n|___ '
                                      '\\\n ___) | \n|____/')
        assert self.five.stars() == '*****'
        assert self.five.bool() == True
        assert self.five.elements() == ["Metal", "Wood", "Water", "Fire",
                                        "Earth"]
        assert self.five.senses() == ["Sight", "Hearing", "Taste", "Smell",
                                      "Touch"]
        assert self.five.circled() == '⑤'
        assert self.five.flipside() == 'ϛ'
        assert self.five.lcd() == ' ---\n|\n --\n   |\n---'

    def test_multilingual_five(self):
        assert self.five.afrikaans() == 'vyf'
        assert self.five.armenian() == 'հինգ'
        assert self.five.arabic() == 'خمسة'
        assert self.five.azerbaijani() == 'beş'
        assert self.five.basque() == 'bost'
        assert self.five.belarusian() == 'пяць'
        assert self.five.bosnian() == 'pet'
        assert self.five.bulgarian() == 'пет'
        assert self.five.canadian() == 'five eh'
        assert self.five.catalan() == 'cinc'
        assert self.five.chinese() == '五'
        assert self.five.chinese('pinyin') == 'wǔ'
        assert self.five.chinese('financial') == '伍'
        assert self.five.choctaw() == 'tahlapi'
        assert self.five.croatian() == 'pet'
        assert self.five.czech() == 'pět'
        assert self.five.danish() == 'fem'
        assert self.five.dovah() == 'hen'
        assert self.five.dutch() == 'vijf'
        assert self.five.elvish() == 'lempe'
        assert self.five.english() == 'five'
        assert self.five.estonian() == 'viis'
        assert self.five.finnish() == 'viisi'
        assert self.five.french() == 'cinq'
        assert self.five.georgian() == 'ხუთი'
        assert self.five.german() == 'fünf'
        assert self.five.greek() == 'πέντε'
        assert self.five.hebrew() == 'חמש'
        assert self.five.hindi() == 'पांच'
        assert self.five.hungarian() == 'öt'
        assert self.five.icelandic() == 'fimm'
        assert self.five.indonesian() == 'lima'
        assert self.five.irish() == 'cúig'
        assert self.five.italian() == 'cinque'
        assert self.five.japanese() == '五'
        assert self.five.kannada() == 'ಐದು'
        assert self.five.klingon() == 'vagh'
        assert self.five.korean() == '오'
        assert self.five.latin() == 'quinque'
        assert self.five.latvian() == 'pieci'
        assert self.five.lithuanian() == 'penki'
        assert self.five.lojban() == 'mu'
        assert self.five.manx() == 'queig'
        assert self.five.mongolian() == 'таван'
        assert self.five.norwegian() == 'fem'
        assert self.five.persian() == 'پنج'
        assert self.five.piglatin() == 'ivefay'
        assert self.five.polish() == 'pięć'
        assert self.five.portuguese() == 'cinco'
        assert self.five.punjabi() == 'ਪੰਜ'
        assert self.five.romanian() == 'cinci'
        assert self.five.russian() == 'пять'
        assert self.five.serbian() == 'pet'
        assert self.five.slovakian() == 'päť'
        assert self.five.slovenian() == 'pet'
        assert self.five.spanish() == 'cinco'
        assert self.five.swedish() == 'fem'
        assert self.five.tagalog() == 'lima'
        assert self.five.tamil() == 'ஐந்து'
        assert self.five.telugu() == 'ఐదు'
        assert self.five.turkish() == 'beş'
        assert self.five.thai() == 'ห้า'
        assert self.five.xhosa() == 'zintlanu'
        assert self.five.ukrainian() == 'п’ять'
        assert self.five.urdu() == 'پانچ'
        assert self.five.vietnamese() == 'năm'
        assert self.five.welsh() == 'pump'

    def test_redix_of_five(self):
        assert self.five.binary() == '101'
        assert self.five.octal() == '5'
        assert self.five.hex() == '5'

    def test_map_and_reduce_of_five(self):
        assert self.five.map([1, 2, 3]) == [5, 5, 5]
        assert self.five.reduce([1, 2, 3]) == 5

    def test_novelty_five(self):
        assert self.five.jackson() == ['Jackie', 'Tito', 'Jermaine', 'Marlon',
                                       'Michael']
        assert self.five.fab() == ['Juwan Howard', 'Ray Jackson', 'Jimmy King',
                                   'Jalen Rose', 'Chris Webber']
        assert self.five.famous_five() == ['Julian', 'Dick', 'Anne', 'George',
                                           'Timmy']

    def test_rotation_of_five(self):
        assert self.five.rotate(string.digits) == '5678901234'
        assert self.five.rotate(string.ascii_lowercase) == (
            'fghijklmnopqrstuvwxyzabcde')
        assert self.five.rotate(string.ascii_uppercase) == (
            'FGHIJKLMNOPQRSTUVWXYZABCDE')
        assert self.five.rotate('$_$ -,- @.@?') == '$_$ -,- @.@?'

    def test_unicode_of_five(self):
        assert self.five.oclock() == '🕔'


five/five.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import absolute_import, unicode_literals, division

import math
import time
import string
from functools import reduce


class Five(object):
    # Start of "basic five'.
    def five(self, *args):
        return 5

    def pow(self, num):
        return math.pow(5, num)

    def sqrt(self):
        return math.sqrt(5)
    # End of "basic five".

    # Start of "math".
    def fibonacci(self):
        "The fifth num in the fibonacci sequence."
        return 5

    def factorial(self):
        "The factorial of five: 5!"
        return 120

    def factors(self):
        return [1, 5]

    def is_prime(self):
        return True
    # End of "math".

    def __call__(self):
        return self.five()

    # Start of "different sorts of five".
    def up_high(self):
        return '⁵'

    def down_low(self):
        return '₅'

    def too_slow(self):
        time.sleep(0.555)
        return self.five()

    def roman(self):
        return 'V'

    def morse_code(self):
        return '.....'

    def negative(self):
        return -5

    def loud(self, lang='englist'):
        """Speak loudly! FIVE! Use upper case!"""
        lang_method = getattr(self, lang, None)
        if lang_method:
            return lang_method().upper()
        else:
            return self.english().upper()

    def smooth(self):
        return 'S'

    def figlet(self):
        return ' _____\n| ____|\n| |__\n|___ \\\n ___) | \n|____/'

    def stars(self):
        return '*****'

    def bool(self):
        """Five should be true."""
        return True if 5 else False

    def elements(self):
        """The five basic elements on earth."""
        return ["Metal", "Wood", "Water", "Fire", "Earth"]

    def senses(self):
        """The five senses that provide data for perception."""
        return ["Sight", "Hearing", "Taste", "Smell", "Touch"]

    def circled(self):
        return '⑤'

    def flipside(self):
        """A five on the flipside should be upside down."""
        return 'ϛ'

    def lcd(self):
        return ' ---\n|\n --\n   |\n---'
    # End of "different sorts of five"

    # Start of "kinds of five in many languages",
    # sorted by alphabetic.
    def afrikaans(self):
        return 'vyf'

    def armenian(self):
        return 'հինգ'

    def arabic(self):
        return 'خمسة'

    def azerbaijani(self):
        return 'beş'

    def basque(self):
        return 'bost'

    def belarusian(self):
        return 'пяць'

    def bosnian(self):
        return 'pet'

    def bulgarian(self):
        return 'пет'

    def canadian(self):
        return 'five eh'

    def catalan(self):
        return 'cinc'

    def chinese(self, case=None):
        if case == 'pinyin':
            return 'wǔ'
        elif case == 'financial':
            return '伍'
        else:
            return '五'

    def choctaw(self):
        return 'tahlapi'

    def croatian(self):
        return 'pet'

    def czech(self):
        return 'pět'

    def danish(self):
        return 'fem'

    def dovah(self):
        return 'hen'

    def dutch(self):
        return 'vijf'

    def elvish(self):
        return 'lempe'

    def english(self):
        return 'five'

    def estonian(self):
        return 'viis'

    def finnish(self):
        return 'viisi'

    def french(self):
        return 'cinq'

    def georgian(self):
        return 'ხუთი'

    def german(self):
        return 'fünf'

    def greek(self):
        return 'πέντε'

    def hebrew(self):
        return 'חמש'

    def hindi(self):
        return 'पांच'

    def hungarian(self):
        return 'öt'

    def icelandic(self):
        return 'fimm'

    def indonesian(self):
        return 'lima'

    def irish(self):
        return 'cúig'

    def italian(self):
        return 'cinque'

    def japanese(self):
        return '五'

    def kannada(self):
        return 'ಐದು'

    def klingon(self):
        return 'vagh'

    def korean(self):
        return '오'

    def latin(self):
        return 'quinque'

    def latvian(self):
        return 'pieci'

    def lithuanian(self):
        return 'penki'

    def lojban(self):
        return 'mu'

    def manx(self):
        return 'queig'

    def mongolian(self):
        return 'таван'

    def norwegian(self):
        return 'fem'

    def persian(self):
        return 'پنج'

    def piglatin(self):
        return 'ivefay'

    def polish(self):
        return 'pięć'

    def portuguese(self):
        return 'cinco'

    def punjabi(self):
        return 'ਪੰਜ'

    def romanian(self):
        return 'cinci'

    def russian(self):
        return 'пять'

    def serbian(self):
        return 'pet'

    def slovakian(self):
        return 'päť'

    def slovenian(self):
        return 'pet'

    def spanish(self):
        return 'cinco'

    def swedish(self):
        return 'fem'

    def tagalog(self):
        return 'lima'

    def tamil(self):
        return 'ஐந்து'

    def telugu(self):
        return 'ఐదు'

    def turkish(self):
        return 'beş'

    def thai(self):
        return 'ห้า'

    def ukrainian(self):
        return 'п’ять'

    def xhosa(self):
        return 'zintlanu'

    def urdu(self):
        return 'پانچ'

    def vietnamese(self):
        return 'năm'

    def welsh(self):
        return 'pump'
    # End of "kinds of five in many languages".

    # Start of "different radices of five".
    def binary(self):
        return '101'

    def octal(self):
        return '5'

    def hex(self):
        return '5'
    # End of "different radices of five".

    # Start of "map and reduce".
    def map(self, seq):
        return list(map(self.five, seq))

    def reduce(self, seq):
        return reduce(self.five, seq)
    # End of "map and reduce".

    # Start of "novelty five".
    def fab(self):
        return ['Juwan Howard', 'Ray Jackson', 'Jimmy King', 'Jalen Rose',
                'Chris Webber']

    def jackson(self):
        return ['Jackie', 'Tito', 'Jermaine', 'Marlon', 'Michael']

    def famous_five(self):
        """The Famous Five is the name of a series of children's
        adventure novels by English author Enid Blyton."""
        return ['Julian', 'Dick', 'Anne', 'George', 'Timmy']
    # End of "novelty five".

    def rotate(self, word):
        """Replaced by a letter 5 right shift.
         e.g. a->f, b->g, . -> ."""
        before = string.printable[:62]
        after = ''.join([i[5:] + i[:5] for i in [string.digits,
                                                 string.ascii_lowercase,
                                                 string.ascii_uppercase]])
        table = dict(zip(before, after))
        processed_word = [table[i] if i in before else i for i in word]
        return ''.join(processed_word)

    def oclock(self):
        return '🕔'
